# Exercise 3: Elliptic Curve Cryptography

**COMP842 Applied Blockchains and Cryptocurrencies**

- Name: Sokhour Lay
- Student ID: 25314544

### 1. Generate the ECDSA key pair

In [5]:
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives import serialization, hashes
from cryptography.exceptions import InvalidSignature

# Generate an ECDSA key pair on secp256k1 (same method as the Week 3 tutorial)
private_key = ec.generate_private_key(ec.SECP256K1(), default_backend())
public_key = private_key.public_key()

# Save both keys in PEM format
private_ks = private_key.private_bytes(
    serialization.Encoding.PEM,
    serialization.PrivateFormat.TraditionalOpenSSL,
    serialization.NoEncryption())
public_ks = public_key.public_bytes(
    serialization.Encoding.PEM,
    serialization.PublicFormat.SubjectPublicKeyInfo)

print(private_ks.decode())
print(public_ks.decode())

-----BEGIN EC PRIVATE KEY-----
MHQCAQEEIJKPsUIY6oMGyEa/Kmc3jTrwZ/XRxbDRvaKbm9GIvhfuoAcGBSuBBAAK
oUQDQgAE+t+2K2ByFsdWYzjwXWJgkBDlSx5Q/jWIJ2PT3tyP9qwuuYvtscftRL5J
nEUz8kMylquECsP4dupCdPgD5fjdGA==
-----END EC PRIVATE KEY-----

-----BEGIN PUBLIC KEY-----
MFYwEAYHKoZIzj0CAQYFK4EEAAoDQgAE+t+2K2ByFsdWYzjwXWJgkBDlSx5Q/jWI
J2PT3tyP9qwuuYvtscftRL5JnEUz8kMylquECsP4dupCdPgD5fjdGA==
-----END PUBLIC KEY-----



**Explanation:** an ECDSA key pair is generated on the secp256k1 curve using the `cryptography` library, and both keys are saved in PEM format. The private key is a secret number, and the public key is a point on the curve derived from it.

### 2. Compare key sizes

In [6]:
# Get the raw key values in hexadecimal
numbers = public_key.public_numbers()
x_hex = format(numbers.x, "064x")
y_hex = format(numbers.y, "064x")
uncompressed_hex = "04" + x_hex + y_hex          # 04 + x + y
priv_hex = format(private_key.private_numbers().private_value, "064x")

print("Private key (hex):", priv_hex)
print("Public key  (hex):", uncompressed_hex)
print()
print("Private key:", len(priv_hex)//2, "bytes /", len(priv_hex), "hex characters")
print("Public key :", len(uncompressed_hex)//2, "bytes /", len(uncompressed_hex), "hex characters")

Private key (hex): 928fb14218ea8306c846bf2a67378d3af067f5d1c5b0d1bda29b9bd188be17ee
Public key  (hex): 04fadfb62b607216c7566338f05d62609010e54b1e50fe35882763d3dedc8ff6ac2eb98bedb1c7ed44be499c4533f2433296ab840ac3f876ea4274f803e5f8dd18

Private key: 32 bytes / 64 hex characters
Public key : 65 bytes / 130 hex characters


**Explanation:** the private key is 32 bytes (64 hex characters), while the uncompressed public key is 65 bytes (130 hex characters), because it stores both the x and y coordinates of the curve point together with a one-byte prefix.

### 3. Compress the public key

In [7]:
# Compress the public key by hand: keep x, and use a 02/03 prefix for the parity of y
prefix = "02" if numbers.y % 2 == 0 else "03"
compressed_hex = prefix + x_hex
reduction = (1 - len(compressed_hex) / len(uncompressed_hex)) * 100

print("Public key uncompressed:", uncompressed_hex)
print("Public key compressed  :", compressed_hex)
print("Compressed:", len(compressed_hex)//2, "bytes /", len(compressed_hex), "hex characters")
print(f"Storage reduction: {reduction:.1f}%")

Public key uncompressed: 04fadfb62b607216c7566338f05d62609010e54b1e50fe35882763d3dedc8ff6ac2eb98bedb1c7ed44be499c4533f2433296ab840ac3f876ea4274f803e5f8dd18
Public key compressed  : 02fadfb62b607216c7566338f05d62609010e54b1e50fe35882763d3dedc8ff6ac
Compressed: 33 bytes / 66 hex characters
Storage reduction: 49.2%


**Explanation:** compression keeps the x coordinate and replaces the long y coordinate with a one-byte prefix (02 if y is even, 03 if y is odd). This reduces the public key from 65 bytes to 33 bytes, a reduction of about 49%.

### 4. Sign a message and verify with the compressed key

In [8]:
message = b"Academic record: S2231 completed COMP842"

# Sign the message with the private key
signature = private_key.sign(message, ec.ECDSA(hashes.SHA256()))

# Decompress: rebuild the full public key from the compressed bytes
compressed_bytes = bytes.fromhex(compressed_hex)
recovered_key = ec.EllipticCurvePublicKey.from_encoded_point(ec.SECP256K1(), compressed_bytes)

# Verify the signature using the recovered (compressed) public key
try:
    recovered_key.verify(signature, message, ec.ECDSA(hashes.SHA256()))
    print("Signature valid using the compressed public key?", True)
except InvalidSignature:
    print("Signature valid using the compressed public key?", False)

Signature valid using the compressed public key? True


**Explanation:** the private key signs a short message, and the compressed public key is rebuilt (decompressed) and used to verify the signature. The successful result confirms that the compressed key still works correctly for verification.

### Reflection Questions

#### 1. Why is it possible to compress an elliptic curve public key but not the corresponding private key?

An elliptic curve public key is a point (x, y) on the curve, and because both coordinates must satisfy the curve equation, the y coordinate is determined by x up to two possible values, one even and one odd. Only the x coordinate and a single parity bit are therefore needed to reconstruct the full point, which is why the public key can be compressed to about half of its original size (Narayanan et al., 2016). In this exercise, the public key was compressed by keeping only the x coordinate and adding a one byte prefix, 02 when y is even and 03 when y is odd, which reduced it from 65 bytes to 33 bytes, a reduction of about 49 percent. A private key, in contrast, is simply a randomly chosen scalar with no algebraic structure or internal redundancy, so every bit is essential and nothing about the value can be derived from any other part (Narayanan et al., 2016). The private key is therefore already maximally information dense and cannot be compressed.

#### 2. If every public key stored on a blockchain were compressed, how would this affect blockchain storage requirements, network bandwidth, and transaction efficiency?

Compressing every public key stored on a blockchain would bring clear and compounding benefits across storage, bandwidth, and transaction efficiency. In this exercise, compression reduced each public key from 65 bytes to 33 bytes, a reduction of about 49 percent, so applying it to every key roughly halves the space that public keys occupy on the chain. This matters because blockchains are already very large: by 2019 a full Bitcoin node already had to download and process around 197 GB of chain data (Marsalek et al., 2019), and the chain has continued to grow well beyond that size since, so a saving on every key accumulates into a significant reduction at network scale (Heo et al., 2024). Smaller stored data also means less data to transmit, so compression lowers the bandwidth needed to propagate and synchronise the chain and reduces the time and resources spent on storing and retrieving it (Laouid et al., 2023). Finally, smaller keys make transactions smaller, which lets more transactions fit inside a fixed block size and improves overall throughput (Heo et al., 2024). The only real trade-off is a small decompression cost at verification, because the y coordinate must be recomputed from x, but this calculation is cheap for a standard curve such as secp256k1 and is easily outweighed by the storage and bandwidth savings. For these reasons, public key compression is best understood as a baseline efficiency measure rather than an optional extra, and it is already used by default in networks such as Bitcoin.